circuits/update_state_tree.circom

```diff
@@ -1,106 +1,266 @@
+  signature_verifier.from_y <== state_tree_data[1]; // public key y
+
+  signature_verifier.R8x <== decrypted_command.out[message_length - 4]; // sig R8x
+  signature_verifier.R8y <== decrypted_command.out[message_length - 3]; // sig R8x
+  signature_verifier.S <== decrypted_command.out[message_length - 2]; // sig S
+
+  for (var i = 0; i < message_length - 4; i++) {
+    signature_verifier.preimage[i] <== decrypted_command.out[i];
+  }
+
+  // Calculate new vote credits
+  var new_vote_credits = state_tree_data[3] - (decrypted_command.out[4] * decrypted_command.out[4]);
+
+  // Construct new state tree data (and its hash)
+  var new_state_tree_data = [
+    decrypted_command.out[1],
+    decrypted_command.out[2],
+    new_vote_options_tree.root,
+    new_vote_credits,
+    decrypted_command.out[5]
+  ];
+
+  component new_state_tree_leaf = Hasher(state_tree_data_length);
+  new_state_tree_leaf.key <== 0;
+  for (var i = 0; i < state_tree_data_length; i++) {
+    new_state_tree_leaf.in[i] <== new_state_tree_data[i];
   }
 
-  // **** 3.2 Validate signature against existing public key **** //
-  component signature_verifier = VerifySignature(message_length);
+  // Checks to see if we should update the
+  // random leaf (Note: 0 is true, anything non-zero is false)
+  signal valid_signature;
+  valid_signature <-- signature_verifier.valid == 0 ? 0 : 1;
 
-  signature_verifier.from_x <== existing_public_key[0]; // public key x
-  signature_verifier.from_y <== existing_public_key[1]; // public key y
+  signal sufficient_vote_credits;
+  sufficient_vote_credits <-- new_vote_credits > 0 ? 0 : 1;
 
-  signature_verifier.R8x <== decrypted_data.out[message_length]; // sig R8x
-  signature_verifier.R8y <== decrypted_data.out[message_length + 1]; // sig R8x
-  signature_verifier.S <== decrypted_data.out[message_length + 2]; // sig S
+  signal correct_nonce;
+  correct_nonce <-- decrypted_command.out[5] == (state_tree_data[4] + 1) ? 0 : 1;
 
-  for (var i=0; i < message_length; i++) {
-    signature_verifier.preimage[i] <== decrypted_data.out[i];
+  signal valid_state_leaf_index;
+  valid_state_leaf_index <-- decrypted_command.out[0] <= state_tree_max_leaf_index ? 0 : 1;
+
+  signal valid_vote_options_leaf_index;
+  valid_vote_options_leaf_index <-- decrypted_command.out[3] <= vote_options_max_leaf_index ? 0 : 1;
+
+  signal valid_update;
+  valid_update = valid_signature + sufficient_vote_credits + correct_nonce + valid_state_leaf_index + valid_vote_options_leaf_index;
+
+
+  // If any of the above checks return false
+  // We update the random leaf instead
+
+  // Note: valid_update 0 is true, 1 is false
+  signal selected_state_tree_hash;
+  selected_state_tree_hash <-- valid_update == 0 ? new_state_tree_leaf.hash : random_leaf;
+
+  signal selected_state_tree_path_elements[depth];
+  for (var i = 0; i < depth; i++) {
+    selected_state_tree_path_elements[i] <-- valid_update == 0 ? state_tree_path_elements[i] : random_leaf_path_elements[i];
```

Comment:
```suggestion
    selected_state_tree_path_elements[i] <== valid_update == 0 ? state_tree_path_elements[i] : random_leaf_path_elements[i];
```

Is it possible to use `<==` just in case?